### ISSUE TITLE
`Port` object - services

### ISSUE THREAD

**Issue by 0xdevalias** [ReconJSON/ReconJSON#19]
Skimming through the current `port` object, it states that:

> services - MUST be a list of Service object(s) that describes the service(s) behind this port

But then in the example that follows, the `services` key isn't defined for the first entry.

I expect the intention of this is that if the `services` key is defined, then it must be an array of service objects. At least, this is how I would expect it to be.

I'm not sure that it would make sense to enforce a MUST on the service objects, as there is a potential where we know a port is open but know nothing else about what is behind it.

**Comment by Rhynorater** [comment]
So, perhaps I'm doing this wrong, but I'm using MUST to mean "if this object is defined, then this MUST happen" and then I'm using the * to indicate whether or not certain fields are required in the object structure. In the `Port` object, since `services` isn't required, I don't give an example with it. 

For an example of the implementation of the `services` attribute of `Port`, see the `Service` object. 

If you have any ideas on how that could be done better, please let me know. Clearly it wasn't clear enough for you, so others are probably making a similar mistake. @0xdevalias

**Comment by 0xdevalias** [comment]
I don't believe that usage quite matches the standard usage (from my understanding). I believe this is the correct standard, but I only googled super quick.. https://www.ietf.org/rfc/rfc2119.txt

For example:

We have an object `Foo` which `MAY` exist.

`Foo` objects:
* `MUST` contain a key `Bar` of type String
* `MAY` contain a key `Baz` of type Array of Strings

Looking at the [Port](https://github.com/ReconJSON/ReconJSON/blob/master/README.md#port) section:

> The PORT object SHALL be used to describe the status of the ports on a specific Host.

I think the use of SHALL here isn't needed. Probably just "The PORT object is used to describe the status of the ports on a specific Host."

Then the next section describes the MUST of how it is used:

> The Port object MUST be placed inside of one of the two ISO Model Layer 4 categories of the Host object.

The following section would tell me that attribute values are optional, but MUST NOT be included if they are unknown. Though this would seem to conflict with your later attempt to distinguish required/optional attributes with a `*`. So I would probably remove it (as it adds ambiguity), and handle that distinction using the proper keywords in the dotpoints below..

> If attribute values are not known, they MUST NOT be included unless specified otherwise in the description below.

So to combine all of that together.. I think this is how I would expect that section to look:

> The `port` object is used to describe the status of the ports on a specific `host`. The `port` object MUST be placed inside one of the two ISO Model Layer 4 categories of the `host` object. (See XXHOSTSECTIONYY for more details).
> 
> A `port` object has the following attributes:
> 
> * `type`: MUST be `port`
> * `port`: MUST be an integer between `0` and `65535` representing the `port` this object describes
> * `state`: Represents the 'openness state' of the port. MUST be one of `open`, `filtered`, `closed`. In most cases, `closed` ports SHOULD NOT be included.
> * `services`: MUST be an array of `service` object(s) that describe the service(s) behind this port, or an empty array

I think in cases such as the above, where we include arrays of things, it probably makes more sense to style things like `MUST be an array of X, or an empty array`, rather than not including any value there. That is just my personal opinion though, and am open to others views on it. I feel like it makes things a bit 'easier to read' when inspecting the raw json, and aids in discoverability (rather than me having to personally know/understand the spec and know that there should be a services array here)

I also feel like it might make sense to start collecting some of these 'high level'/stylistic 'rules' in a metadata/style guide type section of this repo. I would see this section as helping to inform spec contributors about the intended style used throughout the spec, and we could start it off really basic and build it up as questions/conflicts arise (eg. Use keywords as defined in `RFC2119`, use single backticks for inline code blocks and triple backticks for standalone code blocks, what casing to use for types (#22), etc)

**Comment by Rhynorater** [comment]
I think I get what you're saying. 

So, I don't think the "MUST" keyword implies that the field is required, just that the proceeding statement is a requirement of the protocol. However, if you like, we could do something like this:
`optionalField`: MAY be included in the object. MUST be XYZ...

Thoughts?


Or perhaps we could do something like this:

The `port` object MUST have the following attributes:
* X
* Y
* Z

The `port` object MAY have the following attributes:
* A 
* B 
* C

Perhaps this one is better?

**Comment by 0xdevalias** [comment]
By my reading of https://www.ietf.org/rfc/rfc2119.txt

> 1. MUST   This word, or the terms "REQUIRED" or "SHALL", mean that the
   definition is an absolute requirement of the specification.

If my description of an attribute (to take an example from above):
> `services`: MUST be an array of `service` object(s) that describe the service(s) behind this port, or an empty array

The "definition" here is:
> an array of `service` object(s) that describe the service(s) behind this port, or an empty array

Sine this is "absolute requirement of the specification", then the `services` key can never be anything other than:
* an array of `service` object(s) that describe the service(s) behind this port
* an empty array

So by my read it absolutely defines that the field is required (as not being there is not a valid option in the definition).

Though I can see your point, and if you want to be super explicit, the must/may section breakdown you have above is probably even clearer. I would have to turn to looking at other specs to see how they handle these sorts of things to know for sure.

eg. https://tools.ietf.org/html/rfc7519#section-4.1.2 seems to include 'use of this claim is OPTIONAL' at the end, as well as using some MUST language within it.

I think without delving too deep into perfect RFC spec land, your compromise seems decent for differentiating required vs optional, is less verbose than the above example, and is far better/more 'speclike' than the current `*` type denotion.
